python/dgl/data/utils.py

```diff
@@ -482,3 +484,198 @@ def add_nodepred_split(dataset, ratio, ntype=None):
+
+    Parameters
+    ----------
+    dataset : DGLDataset or list of :class:`DGLGraph`
+        The dataset to induce structural distributional shift in.
+    part_ratios : list
+        A list of 5 ratio values for training, ID validation, ID test,
+        OOD validation and OOD test parts. The values must sum to 1.0.
+    property_name : str
+        The node property name to split the dataset by. Must be
+        ``'popularity'``, ``'locality'`` or ``'density'``.
+    ascending : bool, optional
+        Whether to sort nodes in the ascending order of a particular
+        node property, so that more shifted OOD nodes have greater values
+        of the computed property (default: True)
+    random_seed : int, optional
+        Random seed to fix for the initial permutation of nodes. It is
+        used to create a random order for the nodes that have the same
+        property values. (default: None)
+
+    Examples
+    --------
+    >>> dataset = dgl.data.AmazonCoBuyComputerDataset()
+    >>> print('in_valid_mask' in dataset[0].ndata)
+    False
+    >>> part_sizes = [0.3, 0.1, 0.1, 0.3, 0.2]
+    >>> property_name = 'popularity'
+    >>> dgl.data.utils.add_node_property_split(dataset, part_sizes, property_name)
+    >>> print('in_valid_mask' in dataset[0].ndata)
+    True
+    """
+
+    assert property_name in [
+        "popularity",
+        "locality",
+        "density",
+    ], "The name of property has to be 'popularity', 'locality', or 'density'"
+
+    assert (
+        len(part_ratios) == 5
+    ), "The list of part ratios must contain 5 values"
+
+    try:
+        import graph_tool as gt
+        from graph_tool import centrality, clustering, generation
+    except ImportError:
+        warnings.warn(
+            "graph-tool is required to compute the node property values: "
+            "https://graph-tool.skewed.de/"
+        )
+
+    user_direction = 1 if ascending else -1
+
+    for idx in range(len(dataset)):
+        graph_dgl = dataset[idx]
+        num_nodes = graph_dgl.num_nodes()
+
+        edge_list = F.stack(graph_dgl.edges(), dim=0)
+        if F.backend_name == "mxnet":
+            edge_list = edge_list.asnumpy().T
+        else:
+            edge_list = edge_list.numpy().T
+
+        graph_gt = gt.Graph()
+        graph_gt.add_vertex(num_nodes)
+        graph_gt.add_edge_list(edge_list)
+        graph_gt.set_directed(False)
+        generation.remove_parallel_edges(graph_gt)
+
+        if property_name == "popularity":
+            default_direction = -1
+            property_values = (
+                user_direction
+                * default_direction
+                * np.array(centrality.pagerank(graph_gt).get_array())
+            )
+
+        if property_name == "locality":
+            default_direction = -1
+            pagerank_values = np.array(
+                centrality.pagerank(graph_gt).get_array()
+            )
+
+            ohe_mask = np.zeros_like(pagerank_values)
```

Comment:
Leave a comment to explain the meaning of `ohe` or use a more readable variable name.